### PR TITLE
Implement proper logging HttpMiddleware

### LINKED ISSUE #71: Implement proper logging HttpMiddleware
There is an `HttpMiddleware.logging` middleware, but it is not completed. It should be rewritten to utilize the [Common Log Format](https://en.wikipedia.org/wiki/Common_Log_Format), as well as be more flexible where it places the log output, possibly using a `ZSink` from ZIO Streams to accept all the log lines.

This ticket will be done when the logging middleware can write to file output and writes in the common log format, with tests for both.

### PR DESCRIPTION
**Pull request by anovakovic01**
This PR resolves #71.

**Note:** Added HTTP version as a String parameter of the `run` method. This should be changed when the HTTP version is modeled.

### CODE REVIEW

File: core/src/main/scala/zio/web/http/HttpMiddleware.scala [hunk 1]
```diff
@@ -76,13 +81,21 @@ object HttpMiddleware {
       )
     )
 
-  val logging: HttpMiddleware[zio.console.Console, Nothing] =
+  def logging(dest: ZManaged[Blocking, IOException, OutputStream]): HttpMiddleware[Blocking with Clock, Exception] =
```
Comment (jdegoes) [inline core/src/main/scala/zio/web/http/HttpMiddleware.scala:84]: What I would do here is maybe accept a `Sink` that takes the input lines (`String`). Then have a default method that creates a sink that logs to a file (obviously have to think about log rotation and such but no need to do that now).

File: core/src/test/scala/zio/web/HttpMiddlewareSpec.scala [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+  // TODO: should read logs from file and assert content
+  def spec =
+    suite("HttpMiddleware")(
+      suite("logging")(
+        testM("write to file") {
+          val dest = HttpMiddleware.fileSink("test.log")
+
+          for {
+            l      <- HttpMiddleware.logging(dest).make
+            result <- l.request.processor(("GET", new URI("http://zio.dev")))
```
Comment (anovakovic01) [inline core/src/test/scala/zio/web/HttpMiddlewareSpec.scala:26]: @jdegoes What is the easiest way to trigger middleware logic from test?

File: core/src/main/scala/zio/web/http/HttpMiddleware.scala [hunk 3]
```diff
@@ -76,16 +81,28 @@ object HttpMiddleware {
...
+        stream = ZStream.fromQueue(queue)
+        _      <- stream.run(sink).fork
+      } yield Middleware(
+        request(HttpRequest.Method.zip(HttpRequest.URI)) { tuple =>
+          for {
+            now <- currentDateTime.orDie
+            _   <- queue.offer(s"- - - [$now] \'${tuple._1} ${tuple._2}\' - -\n")
+          } yield ()
+        },
+        Response.none
```
Comment (anovakovic01) [inline core/src/main/scala/zio/web/http/HttpMiddleware.scala:97]: I still need to somehow connect Response to the logging and add the response code and size.

**Comment by jdegoes** [comment]
@anovakovic01 You need any more direction / help with this?

The other thing I would do is move this to `zio.web.http.middleware`.

Looking good!

**Comment by anovakovic01** [comment]
@jdegoes I'm almost done, just need to finish tests (that will be done as soon as `run` methods are merged to master). I appreciate your help on the issue. Thank you very much!

**Comment by pshemass** [comment]
@jdegoes should this middleware use `zio-logging` or maybe this middleware should be in `zio-logging`?

File: core/src/main/scala/zio/web/http/HttpMiddleware.scala [hunk 4]
```diff
@@ -82,15 +93,47 @@ object HttpMiddleware {
       )
     )
 
-  val logging: HttpMiddleware[zio.console.Console, Nothing] =
+  def logging[R, E](sink: ZSink[R, E, String, Byte, Long]): HttpMiddleware[Clock with R, Nothing] = {
```
Comment (jdegoes) [inline core/src/main/scala/zio/web/http/HttpMiddleware.scala:96]: Let's move this logging middleware into `zio.web.http.middleware`, for organizational purposes.

File: core/src/main/scala/zio/web/http/HttpMiddleware.scala [hunk 5]
```diff
@@ -82,15 +93,47 @@ object HttpMiddleware {
...
+        },
+        Response(
+          HttpResponse.StatusCode.zip(HttpResponse.Header("Content-Length")),
+          (state: String, resp: (Int, String)) => queue.offer(s"$state ${resp._1} ${resp._2}\n").as(Patch.empty)
         )
       )
     )
+  }
+
+  def fileSink(path: String): ZSink[Blocking, Throwable, String, Byte, Long] = {
```
Comment (jdegoes) [inline core/src/main/scala/zio/web/http/HttpMiddleware.scala:125]: This looks like a helper, so we can keep it private.

File: core/src/test/scala/zio/web/middleware/HttpMiddlewareSpec.scala [hunk 6]
```diff
@@ -0,0 +1,115 @@
+package zio.web.middleware
```
Comment (jdegoes) [inline core/src/test/scala/zio/web/middleware/HttpMiddlewareSpec.scala:1]: This should be `zio.web.http.middleware`.

**Review by jdegoes — CHANGES_REQUESTED** [review]
Just a few minor tweaks and then good to merge. Thanks for your work on this!

File: core/src/main/scala/zio/web/http/HttpMiddleware.scala (hunk 5 above)
Comment (anovakovic01) [reply core/src/main/scala/zio/web/http/HttpMiddleware.scala:125]: If I make this private, the end users won't be able to create logging middleware that will write to the file. Should I create `fileLogging` middleware and then make this function private?

Comment (jdegoes) [reply core/src/main/scala/zio/web/http/HttpMiddleware.scala:125]: Yes, that sounds like a good midway point. Thank you!

**Comment by jdegoes** [comment]
@anovakovic01 Thank you. Excellent work!

**Pull request merged**